read-fonts/src/tables/postscript/charstring.rs

```diff
@@ -1,203 +1,644 @@
+pub fn evaluate(
+    charstring_data: &[u8],
+    global_subrs: Index,
+    subrs: Option<Index>,
+    blend_state: Option<BlendState>,
+    sink: &mut impl CommandSink,
+) -> Result<(), Error> {
+    let mut evaluator = Evaluator::new(global_subrs, subrs, blend_state);
+    evaluator.evaluate(charstring_data, Fixed::ZERO, Fixed::ZERO, sink, 0)?;
+    Ok(())
 }
 
-impl From<Operator> for Token {
-    fn from(value: Operator) -> Self {
-        Self::Operator(value)
-    }
+struct Evaluator<'a> {
+    global_subrs: Index<'a>,
+    subrs: Option<Index<'a>>,
+    blend_state: Option<BlendState<'a>>,
+    is_open: bool,
+    have_read_width: bool,
+    stem_count: usize,
+    stack: Stack,
 }
 
-impl<T> From<T> for Token
-where
-    T: Into<Number>,
-{
-    fn from(value: T) -> Self {
-        Self::Operand(value.into())
+impl<'a> Evaluator<'a> {
+    fn new(
+        global_subrs: Index<'a>,
+        subrs: Option<Index<'a>>,
+        blend_state: Option<BlendState<'a>>,
+    ) -> Self {
+        Self {
+            global_subrs,
+            subrs,
+            blend_state,
+            is_open: false,
+            have_read_width: false,
+            stem_count: 0,
+            stack: Stack::new(),
+        }
     }
-}
 
-/// Given a byte slice containing charstring data, returns an iterator yielding
-/// raw operands and operators.
-///
-/// This does not perform any evaluation on the charstring.
-pub fn tokens(charstring_data: &[u8]) -> impl Iterator<Item = Result<Token, Error>> + '_ + Clone {
-    let mut cursor = crate::FontData::new(charstring_data).cursor();
-    std::iter::from_fn(move || {
-        if cursor.remaining_bytes() == 0 {
-            None
-        } else {
-            Some(parse_token(&mut cursor))
+    fn evaluate(
+        &mut self,
+        charstring_data: &[u8],
+        mut x: Fixed,
+        mut y: Fixed,
+        sink: &mut impl CommandSink,
+        nesting_depth: u32,
+    ) -> Result<(Fixed, Fixed), Error> {
+        if nesting_depth > NESTING_DEPTH_LIMIT {
+            return Err(Error::CharstringNestingDepthLimitExceeded);
         }
-    })
+        use ops::*;
+        let mut cursor = crate::FontData::new(charstring_data).cursor();
+        while cursor.remaining_bytes() != 0 {
+            let op = cursor.read::<u8>()?;
+            if op == ESCAPE {
+                let two_byte_op = cursor.read::<u8>()?;
+                // The following "flex" operators are intended to emit
+                // either two curves or a straight line depending on
+                // a "flex depth" parameter and the distance from the
+                // joining point to the chord connecting the two
+                // end points. In practice, we just emit the two curves,
+                // following FreeType:
+                // <https://gitlab.freedesktop.org/freetype/freetype/-/blob/80a507a6b8e3d2906ad2c8ba69329bd2fb2a85ef/src/psaux/psintrp.c#L335>
+                match two_byte_op {
+                    HFLEX => {
```

Comment:
Request links to the specific page of documentation, e.g. https://adobe-type-tools.github.io/font-tech-notes/pdfs/5177.Type2.pdf#page=19, at the top of each of these. 

It's a hard to get from the docs to our stream of something = something + something. For example, hflex says `dx1 dx2 dy2 dx3 dx4 dx5 dx6 `. OK, so we start at (x,y) and then the next point is defined only by dx1 so we don't update y, etc. If our sink was friendly to points we could encode this much more directly. For example, HFLEX would become something like:

```rust
let rdr = PointReader::new(&self.stack, x, y, 0);
sink.curve_to(rdr.dx(), rd.dxdy(), rdr.dx());

// PointReader apis:
// dx(): read a point that has only a dx from stack, added to curr x/y
// dxdy(): has x and y from stack
// etc
```

Reply:
(links to specific pages added)

I want to like this one too, but I'm not sure this can be captured with the generality you're assuming. There are a lot of little exceptions to the "delta streaming" model across the operators. In your example, hflex requires resetting y to the initial value at dy5 which would require special processing.

I don't mind playing with something like this, but I'd like to not block this PR on it.

Reply:
Per f2f I still think we're obscuring the logic with details and should at least try doing it in a clearer way. Fine to be in a followon. Tests can ensure we match FT.

Reply:
Will explore this in another PR.